### ISSUE TITLE
MT7603 - persistent issues

### ISSUE THREAD

**Issue by CR-Ryan** [openwrt/mt76#167]
After pulling the latest commit (ca5cc9a), we are still seeing instability on 7603

Behavior is _almost_ identical to what was described [in this post](https://github.com/openwrt/mt76/issues/149#issuecomment-369681365)

The device I am currently testing with is an LG v20,  uses a dual band 11ac chipset.  We have seen this same behavior across many other devices though - it's not just this one. 

After initially connecting, the device remains connected for about 1 - 10 minutes. During which browsing and speedtests work fine. Eventually, the wireless connection gets "stuck".  The device shows as connected to the network, but it cannot ping the AP, and the AP can't ping it.  Obviously browsing doesn't work in this state.  Disconnecting and reconnecting will typically fix this temporarily. Also, setting htmode to 'none'  appears to prevent the issue entirely - but this is not a great fix. 

There is interesting behavior showing in the station dumps. When in this "stuck" state, the rx bitrate will continuously jump between 24Mbits and 144Mbps. The two station dumps below were taken during a speedtest, 1 second apart, and displays this behavior:

root@ph:/etc/config# iw dev wlan0 station dump
Station a8:b8:6e:7d:06:fe (on wlan0)
        inactive time:  110 ms
        rx bytes:       1102876
        rx packets:     11322
        tx bytes:       26489082
        tx packets:     9515
        tx retries:     333
        tx failed:      9
        rx drop misc:   7
        signal:         -46 [-47, -46] dBm
        signal avg:     -47 [-49, -47] dBm
        tx bitrate:     144.4 MBit/s MCS 15 short GI
        rx bitrate:     24.0 MBit/s
        expected throughput:    46.875Mbps
        authorized:     yes
        authenticated:  yes
        associated:     yes
        preamble:       short
        WMM/WME:        yes
        MFP:            no
        TDLS peer:      no
        DTIM period:    2
        beacon interval:100
        short preamble: yes
        short slot time:yes
        connected time: 134 seconds

root@ph:/etc/config# iw dev wlan0 station dump
Station a8:b8:6e:7d:06:fe (on wlan0)
        inactive time:  60 ms
        rx bytes:       1102964
        rx packets:     11324
        tx bytes:       26489144
        tx packets:     9516
        tx retries:     333
        tx failed:      9
        rx drop misc:   7
        signal:         -53 [-54, -53] dBm
        signal avg:     -48 [-49, -48] dBm
        tx bitrate:     144.4 MBit/s MCS 15 short GI
        rx bitrate:     144.4 MBit/s MCS 15 short GI
        expected throughput:    46.875Mbps
        authorized:     yes
        authenticated:  yes
        associated:     yes
        preamble:       short
        WMM/WME:        yes
        MFP:            no
        TDLS peer:      no
        DTIM period:    2
        beacon interval:100
        short preamble: yes
        short slot time:yes
        connected time: 135 seconds

This is in contrast to when it is working well, where the rx bitrate stays on 144Mbps continuously. Here is a station dump from when it was working fine:

Station a8:b8:6e:7d:06:fe (on wlan0)
        inactive time:  10 ms
        rx bytes:       1085972
        rx packets:     11015
        tx bytes:       26449459
        tx packets:     9430
        tx retries:     253
        tx failed:      5
        rx drop misc:   3
        signal:         -49 [-49, -51] dBm
        signal avg:     -47 [-47, -50] dBm
        tx bitrate:     144.4 MBit/s MCS 15 short GI
        rx bitrate:     144.4 MBit/s MCS 15 short GI
        expected throughput:    87.158Mbps
        authorized:     yes
        authenticated:  yes
        associated:     yes
        preamble:       short
        WMM/WME:        yes
        MFP:            no
        TDLS peer:      no
        DTIM period:    2
        beacon interval:100
        short preamble: yes
        short slot time:yes
        connected time: 22 seconds

Here is our wireless config:

config wifi-device 'radio0'
        option type 'mac80211'
        option hwmode '11g'
        option path 'pci0000:00/0000:00:01.0/0000:02:00.0'
        option disabled '0'
        option channel '6'
        option country 'US'
        option txpower '23'
        option noscan '1'
        option htmode 'HT40'
        option log_level '1'

config wifi-iface 'default_radio0'
        option device 'radio0'
        option network 'lan'
        option mode 'ap'
        option hidden '0'
        option disassoc_low_ack '0'
        option ssid 'CleanRouter'
        option encryption 'psk2'
        option key '3453453456'

config wifi-device 'radio1'
        option type 'mac80211'
        option hwmode '11a'
        option path 'pci0000:00/0000:00:00.0/0000:01:00.0'
        option htmode 'VHT80'
        option channel '149'
        option country 'US'
        option txpower '30'
        option noscan '1'
        option disabled '0'
        option log_level '1'

config wifi-iface 'default_radio1'
        option device 'radio1'
        option hidden '0'
        option encryption 'psk2'
        option network 'lan'
        option ssid 'CleanRouter5GHz'
        option disassoc_low_ack '0'
        option mode 'ap'
        option key '3453453456'

Dmesg and logread look fine - no errors present. This is true even when the connection gets "stuck".

I can reproduce this reliably - I am happy to provide more info if needed. 

Thanks!

**Comment by CR-Ryan** [comment]
Update:  Issues still happening on e2eedc9. There is a definite improvement though - it is self-recovering now. Typically, it will recover in about ~30 seconds.  Just like before, station dump shows odd behavior during the connection issues. The three dumps below were taken consecutively over 3 seconds, and you can see the rx bitrate bounce around. 

root@ph:/etc/config# iw dev wlan0 station dump
Station a8:b8:6e:7d:06:fe (on wlan0)
        inactive time:  30 ms
        rx bytes:       9034283
        rx packets:     98952
        tx bytes:       266642551
        tx packets:     98723
        tx retries:     6356
        tx failed:      117
        rx drop misc:   63
        signal:         -39 [-39, -41] dBm
        signal avg:     -38 [-38, -41] dBm
        tx bitrate:     130.0 MBit/s MCS 15
        rx bitrate:     24.0 MBit/s
        expected throughput:    33.507Mbps
        authorized:     yes
        authenticated:  yes
        associated:     yes
        preamble:       short
        WMM/WME:        yes
        MFP:            no
        TDLS peer:      no
        DTIM period:    2
        beacon interval:100
        short preamble: yes
        short slot time:yes
        connected time: 2224 seconds
root@ph:/etc/config# iw dev wlan0 station dump
Station a8:b8:6e:7d:06:fe (on wlan0)
        inactive time:  40 ms
        rx bytes:       9036852
        rx packets:     98973
        tx bytes:       266650403
        tx packets:     98733
        tx retries:     6358
        tx failed:      117
        rx drop misc:   63
        signal:         -38 [-38, -39] dBm
        signal avg:     -38 [-38, -39] dBm
        tx bitrate:     130.0 MBit/s MCS 15
        rx bitrate:     130.0 MBit/s MCS 15
        expected throughput:    37.994Mbps
        authorized:     yes
        authenticated:  yes
        associated:     yes
        preamble:       short
        WMM/WME:        yes
        MFP:            no
        TDLS peer:      no
        DTIM period:    2
        beacon interval:100
        short preamble: yes
        short slot time:yes
        connected time: 2225 seconds
root@ph:/etc/config# iw dev wlan0 station dump
Station a8:b8:6e:7d:06:fe (on wlan0)
        inactive time:  180 ms
        rx bytes:       9037100
        rx packets:     98981
        tx bytes:       266650497
        tx packets:     98734
        tx retries:     6358
        tx failed:      117
        rx drop misc:   63
        signal:         -40 [-40, -40] dBm
        signal avg:     -39 [-39, -39] dBm
        tx bitrate:     130.0 MBit/s MCS 15
        rx bitrate:     24.0 MBit/s
        expected throughput:    40.191Mbps
        authorized:     yes
        authenticated:  yes
        associated:     yes
        preamble:       short
        WMM/WME:        yes
        MFP:            no
        TDLS peer:      no
        DTIM period:    2
        beacon interval:100
        short preamble: yes
        short slot time:yes
        connected time: 2226 seconds

**Comment by Tom-Brouwer** [comment]
I'm seeing the same behaviour as CR-RYAN, using a Xiaomi Router 3G. Some time after connecting (few minutes) loading web sites becomes very slow, and eventually none of them load anymore.

Weirdly enough some existing connections appear to keep functioning during this behavious. E.g. I can access LUCI, and if I start streaming Live TV right after connecting, I can continue streaming without a problem. Like with CR-RYAN, setting HTMODE to NONE, effectively disabling Wireless N, gets rid of the problem. Also I experience none of these problems on my 5GHZ (MT7612). The Wireless config for 2.4Ghz, that gives me the problems is as follows:

```
config wifi-device 'radio0'
        option type 'mac80211'
        option hwmode '11g'
        option path 'pci0000:00/0000:00:00.0/0000:01:00.0'
        option htmode 'HT20'
        option country 'NL'
        option channel '13'
        option log_level '4'
        option txpower '16'
        option legacy_rates '0'

config wifi-iface 'default_radio0'
        option device 'radio0'
        option network 'lan'
        option mode 'ap'
        option encryption 'psk2+ccmp'
        option key 'WIFIPW'
        option ssid 'WIFINAME'
```

**Comment by MaFeSa** [comment]
I have updated to OpenWrt SNAPSHOT r6772 and now I have the same problem with iPad connected to 2.4 GHz.
When it happens in the log I find "disconnected due to excessive missing ACKs", device shows as connected to the network but it cannot ping others devices.
I solved momentarily with `option disassoc_low_ack 0`.

**Comment by MaFeSa** [comment]
As not said, option disassoc_low_ack 0 did not solve the problem.

**Comment by CR-Ryan** [comment]
@MaFeSa -  We are seeing the same thing -  using the disassoc_low_ack 0  option doesn't fix the issue for us either.

**Comment by MaFeSa** [comment]
@CR-Ryan I do not know if it's a coincidence, maybe nbd knows why, but I have updated to r6787 and set option lecacy to 0. 
For now I have no problems for 22 hours with 10 clients connected.

**Comment by CR-Ryan** [comment]
@MaFeSa - I just tried "option legacy_rates '0' (and "1"), and neither seemed to fix the issue for me.  I hadn't heard of Flow Offload before, and couldn't find anything on the wiki about it.  Is that a wireless option?

**Comment by Tom-Brouwer** [comment]
@CR-Ryan I think he means this: https://forum.lede-project.org/t/xiaomi-wifi-router-3g/5377/696 . Its a firewall feature. I continue to have the problems, whether its on or off. If I have time, I'll try it in combination with legacy_rates 0 soon, to see if it has any effect for me...

**Comment by CR-Ryan** [comment]
@Tom-Brouwer Thanks for that - it looks like that feature is off by default, but I set it to 0 just to be safe. Made no difference, issues persist.  I just noticed the config you posted was missing this option:

option disassoc_low_ack '0'

Have you tried running with that set to zero?

**Comment by MaFeSa** [comment]
> For now I have no problems for 22 hours with 10 clients connected.
This morning router reboot itself and now is very unstable, 2.4 GHz it's not usable.

> Have you tried running with that set to zero?

@CR-Ryan @Tom-Brouwer 

Do not waste time, it's useless.

Does anyone remember when the problems started?
With which commit?

**Comment by slthomason** [comment]
We haven't been able to use it on ht40 for quite some time

**Comment by MaFeSa** [comment]
I'll try to figure out where the problem is, I start testing the commits since April 12th.

**Comment by Tom-Brouwer** [comment]
@CR-Ryan I'm not sure, but I think I tried option disassoc_low_ack '0' before. I'll try to test that this weekend.

@slthomason For me, this issue also occurs with HTMODE set to 20...

**Comment by MaFeSa** [comment]
I tried HT20 and HT40 with no results, the only way for a stable 2.4 GHz connection is to set up "Lecacy" with LuCI.

My config for 2.4

```
config wifi-device 'radio0'
	option type 'mac80211'
	option hwmode '11g'
	option path 'pci0000:00/0000:00:00.0/0000:01:00.0'
	option channel '1'
	option country '00'
	option legacy_rates '1'
```

**Comment by slthomason** [comment]
@CR-Ryan and I have found the same thing.  It has been this way for some
time and this latest release is the first to make it better, but still not
totally working.

**Comment by CR-Ryan** [comment]
@Tom-Brouwer  Yeah, HT20 has the same issues. We have been fighting these issues for over a year now.

@MaFeSa  The legacy option didn't do anything for us, but setting the HTmode to none does fix it.

**Comment by MaFeSa** [comment]
@CR-Ryan 
Sorry I explained myself badly...
The legacy option isn't `option legacy 0` but is LuCI option that set HTmode to none
![legacy](https://user-images.githubusercontent.com/1554203/39643288-9b47af32-4fd4-11e8-8832-b3c36c06d122.PNG)

**Comment by MaFeSa** [comment]
> We have been fighting these issues for over a year now.

I have this problem only from April 12th.

**Comment by CR-Ryan** [comment]
@MaFeSa  Ok - thx for clarifying on the legacy option

We first started using MT7603 hardware in April of last year, and had problems from day 1. The HTmodes didn't work correctly back then either. It is definitely better now, but it is still not 100% stable on 7603.  We also tested the stock firmware provided by the manufacturer of our equipment, and the drivers they are using work fine. The issue is they are on a really old kernel, so we can't use their firmware.

**Comment by slthomason** [comment]
@nbd168 - see my comments in https://github.com/openwrt/mt76/issues/152.  But I believe these are the same issues.  Can we implement the same checks and behavior for mt7603?

**Comment by jsantala** [comment]
Yes, I'm also using mt7603e (I think) on VoCore2 and I have packet loss issues (using only adhoc), so if the fix for #152 by @nbd163 also affects mt7603 it would be great news!

**Comment by nbd168** [comment]
I ran some tests, MT7603 does not have the same issue

**Comment by slthomason** [comment]
Hrm.  It very very similar around devices coming out of power save mode.
What do you suggest for this one?  What more data do you want to see?

**Comment by nbd168** [comment]
You could make a pcap of the issue in action, with a separate device for monitoring

**Comment by slthomason** [comment]
Thanks for the suggestion.  Will do.

**Comment by CR-Ryan** [comment]
@nbd168  - Here is our first stab at a packet capture. The tcpdump below was taken while a device was in a "bad state". It was connected to the AP, but couldn't browse or ping the AP.

tcpdump -vv host android-7638b9f93e1c93d0 -i wlan0 > /etc/tcpdump

13:25:49.688688 ARP, Ethernet (len 6), IPv4 (len 4), Reply <the router> is-at 78:a3:51:2d:a1:26 (oui Unknown), length 28
13:25:50.779824 ARP, Ethernet (len 6), IPv4 (len 4), Request who-has <the router> tell android-7638b9f93e1c93d0.lan, length 28
13:25:50.779965 ARP, Ethernet (len 6), IPv4 (len 4), Request who-has <the router> tell android-7638b9f93e1c93d0.lan, length 28
13:25:50.780138 ARP, Ethernet (len 6), IPv4 (len 4), Reply <the router> is-at 78:a3:51:2d:a1:26 (oui Unknown), length 28
13:25:51.698146 ARP, Ethernet (len 6), IPv4 (len 4), Request who-has <the router> tell android-7638b9f93e1c93d0.lan, length 28
13:25:51.698265 ARP, Ethernet (len 6), IPv4 (len 4), Request who-has <the router> tell android-7638b9f93e1c93d0.lan, length 28
13:25:51.698455 ARP, Ethernet (len 6), IPv4 (len 4), Reply <the router> is-at 78:a3:51:2d:a1:26 (oui Unknown), length 28
13:25:52.697837 ARP, Ethernet (len 6), IPv4 (len 4), Request who-has <the router> tell android-7638b9f93e1c93d0.lan, length 28
(This output repeated itself several times over)

I only know of two ways to capture packets: tcpdump and wireshark.  Is there a better way for us to do this? I know wireshark can capture radio layer packets, but that requires "monitor mode", which I have not been able to enable. I don't think my devices wireless driver supports it. 

Please let us know if there is any other info we can grab for you - we are anxious to fix this - we have thousands of customers impacted by this bug.  We can _easily_ reproduce this when using any htmode that is not "NONE". (So HT20/40)   We are totally willing to spin up remote access to both the router AND the affected device during one of these issues if that would help you. 

We know you are busy, and so we appreciate any time you can spend on this. In fact, we would be more than happy to buy you pizza + a sixer of your favorite beer if you can fix this. Seriously. : )

**Comment by nbd168** [comment]
Only real monitor mode captures are helpful for debugging this. The output should be a real .pcap file.

By the way, I've pushed some more fixes for MT7603 and MT7628, please test the latest version from mt76.git (not pushed to OpenWrt yet).

**Comment by Rising-Sun** [comment]
The problem is persistent with driver version d87e4b025286eedc8acffd1af37da695713a9279 and 792dbe02613299f67e92815f0b2b7584dc951ffd. 
Using Xiaomi mir3g as router and iPhone 7 as client.

**Comment by intriguedlife** [comment]
Hello,

I also experience similar problems with MT7603. The issues start immediately after connecting. As has been mentioned; these problems go away when setting htmode to none. I am using a Xiaomi 3G router.

I have captured WLAN frames using my WLAN card in monitor mode:

mt7603.pcapng is a capture when htmode is set to HT40 and running a speedtest. Initially there is no internet connection, after disconnection from the AP and reconnecting a connection is possible but the speeds fluctuate and then go down to no connectivity yet again.

mt7603_htmode_none.pcapng is a capture when htmode is set to none and running the same speedtest. I experienced no connectivity or stability problems at all.

There are a lot of "Null function" frames present in both captures. I am no wifi expert, but this looks weird to me.

(you can use the following display filter with wireshark to hide view obstructing beacon frames:  wlan.fc.type_subtype != 0x08 )

[captures.zip](https://github.com/openwrt/mt76/files/2020675/captures.zip)

Hopefully these captures help to solve these issues.

**Comment by jsantala** [comment]
VoCore2 with mt7603e running OpenWrt SNAPSHOT, r7050-9c409cb:
--- 192.168.12.1 ping statistics ---
172 packets transmitted, 30 packets received, 82% packet loss
round-trip min/avg/max = 55134.079/62646.031/67986.789 ms
Nothing in the logs as far as I can tell. The hosts are right next to each other. The ping will at first drop all packets, but eventually start getting some packets through with huge round-trip values.

**Comment by jsantala** [comment]
Ah, ok, it seems to happen with other wifi too, tried with RT5370 against RTL8192CU, at first anyway:
--- 192.168.13.1 ping statistics ---
85 packets transmitted, 40 packets received, 52% packet loss
round-trip min/avg/max = 2950.277/4330.621/5806.895 ms

But wait, there's more, after I took mt7602e interfaces down (ifconfig down wlan0) and waited a bit, the results on the other usb wifi devices got better, at least part of the time:
--- 192.168.13.1 ping statistics ---
115 packets transmitted, 107 packets received, 6% packet loss
round-trip min/avg/max = 1.689/657.916/2421.064 ms

The RT5270 host had some of this in the logs though:
Mon May 28 08:42:08 2018 kern.warn kernel: [  607.710597] ieee80211 phy1: rt2800usb_txdone: Warning - Data pending for entry 7 in queue 2

So I swapped the RT5270 for a RTL8188CUS and it seems to talk to RTL8192CU quite ok right away:
--- 192.168.13.1 ping statistics ---
101 packets transmitted, 101 packets received, 0% packet loss
round-trip min/avg/max = 2.281/115.664/385.810 ms

Then going back to the mt7603e to mt7603e I get very little again.

All devices being tested at the same adhoc network:
phy#1
	Interface wlan1
		ifindex 5
		wdev 0x100000002
		addr 00:13:ef:50:13:a0
		ssid vad-B9gRYA
		type IBSS
		channel 11 (2462 MHz), width: 20 MHz (no HT), center1: 2462 MHz
		txpower 20.00 dBm
phy#0
	Interface wlan0
		ifindex 6
		wdev 0x2
		addr b8:d8:12:67:68:19
		ssid vad-B9gRYA
		type IBSS
		channel 11 (2462 MHz), width: 20 MHz (no HT), center1: 2462 MHz

I also have batman set up among all the interfaces, but for these test I used the interfaces directly.

**Comment by CR-Ryan** [comment]
Update:
For the last week we have been having issues [described here.](https://github.com/openwrt/mt76/issues/173)

Fortunately, a fix was pushed earlier today and wireless is working correctly again. However, we are still seeing instability on 7603, when using any htmodes.   This is on the latest commit.

**Comment by jsantala** [comment]
I also confirm that on openwrt commit 6c81c27 things are back to "normal" - mt7603e wifi works somewhat, but not enough to maintain a stable ssh connection, for example, even with htmode NONE. I run IBSS only with batman on top. If I switch to known-good USB sticks everything works ok while keeping everything else the same.

**Comment by WBINVD** [comment]
I find the output of tcpdump on wlan0 interesting. When the wireless interface becomes unstable, EAPOL and DHCP work fine, but mDNS and ARP only work in one direction - from the wireless client to the rest of the network.

The ARP respsonses from the LAN network do not show up on the wlan0 tcpdump but do on br-lan. Something is dropping the ARP traffic, possibly before the wireless driver can process it.

I tested this on an MT7620 which I believe uses the rt2800pci driver and not mt76. Setting htmode to none didn't help in my case.

**Comment by lukasz1992** [comment]
I found out, that IEEE80211_TX_STAT_ACK can be set even with MT_TXS0_ACK_TIMEOUT set.

I don't know if repairing it could help, but could you try with replacing

if (!fixed_rate && !ack_timeout)
			sta->ampdu_acked++;
		info->flags |= IEEE80211_TX_STAT_ACK;

With

if (!fixed_rate && !ack_timeout) {
			sta->ampdu_acked++;
		info->flags |= IEEE80211_TX_STAT_ACK;}

?

**Comment by jsantala** [comment]
Current state at openwrt 7590c3c is that if I have just one VoCore2 mt7603e device (adhoc with batman on top) things seem to be quite ok, at least for a while. However, if I add another mt7603e device things get hairy. New device won't work at all and the one next to it that was ok just a moment ago goes bad as well. I can have many devices in the same mesh with different chips, but only one mt7603e at a time. Don't know if this helps at all, but that's my observation as of 7590c3c.

**Comment by jsantala** [comment]
This is what happens when I have a ping against 8.8.8.8 and a second mt7603e device joins the same adhoc+batman network:
64 bytes from 8.8.8.8: seq=24 ttl=56 time=11.771 ms
64 bytes from 8.8.8.8: seq=25 ttl=56 time=11.807 ms
64 bytes from 8.8.8.8: seq=26 ttl=56 time=14.153 ms
64 bytes from 8.8.8.8: seq=27 ttl=56 time=39.726 ms
64 bytes from 8.8.8.8: seq=28 ttl=56 time=1080.775 ms
64 bytes from 8.8.8.8: seq=29 ttl=56 time=2680.045 ms
64 bytes from 8.8.8.8: seq=30 ttl=56 time=3522.285 ms
64 bytes from 8.8.8.8: seq=31 ttl=56 time=4724.123 ms
64 bytes from 8.8.8.8: seq=32 ttl=56 time=5309.847 ms

**Comment by laoshaw** [comment]
from my study into mt7603en it seems this is not purely a software/driver issue, but a hardware one, in short, this 2.4Ghz chip itself is not robust, a router replacement(that does not use mt7603en for 2.4Ghz) is the only dependable fix.

**Comment by slthomason** [comment]
No we have gotten it much more stable and are continuing to make
improvements on it.

**Comment by nbd168** [comment]
Which git revision are you running?

**Comment by laoshaw** [comment]
For openwrt 18.06 git:
commit 159a52e1c2d0889bbb137c42df0062e7df24cac3
Author: Giuseppe Lippolis <giu.lippolis@gmail.com>
Date:   Sun Aug 26 10:52:27 2018 +0200
    comgt: increase timeout on runcommands

for mt76:
commit 14580aaf81c692b0f54b4e7aa003f20eeb8705f6
Author: Felix Fietkau <nbd@nbd.name>
Date:   Wed Aug 22 12:31:55 2018 +0200

**Comment by laoshaw** [comment]
I draw a live graph to monitor /proc/net/wireless and ping at the PC in parallel, in the background I download some data from the router constantly to generate the traffic for my testing.

observation so far:

'ping -c 1' will fail to reach the router randomly. Sometimes ping will fail in a row while wifi signal stays active(network is gone, but not the wifi signal). In fact, when ping fails(i.e. no network connection), the wifi signal actually gets better(from 35 to 45 on link quality). Network connection will recover in a few minutes all by itself.

Will update this late today while the testing is still running.

**Comment by nbd168** [comment]
Please test the latest version from OpenWrt master or the 18.06 branch.

**Comment by laoshaw** [comment]
with brcmsmac which had not even a single issue for weeks with a tplink C7 running 18.06, I just upgraded the newest 18.06 build to a zbt1326 and connected via the same brcmsmac STA, the network(not the wireless) is gone after 20 minutes(downloading traffic at the background via 2.4Ghz), and it does not recover(for another 20 minutes).  the wifi speed is 5Mbps with wget for a 50MB image file.

on another ath9k STA laptop now it's running 40 minutes and the connection stays solid, with downloads at 24Mbps via the same 2.4Ghz(channel 11 which is not crowded here), everything else is the same.

In short, zbt1326 still is not robust, especially when I used it with brcmsmac STA, it is probably worse now, as in previous tests I did not lose network this quickly and if it's lost, most of the time it can recover.

**Comment by nbd168** [comment]
I need to know, which of the last commits made it worse for you. Please try this:

Edit package/kernel/mt76/Makefile, set PKG_SOURCE_VERSION to these values, one by one, in order, and run make (no make clean required) and test again with the brcmsmac sta after each change:
`27af7a570f8eb71b66e98dab8e5a0f6683f292c1`
`497c30431b8635b6865ef80d72cc89784257c9ca`
`6e1898d60a780c9d89dff9cbb3569f267db13e21`
`980c60666eb57daa144d8725712ba303e876e677`
`7daf9621ada28638831beca3073ee3e2b8e6609e`

Please stop at the first hash value that makes it work again, and let me know which one that was.

Thanks

**Comment by laoshaw** [comment]
run each for 2+ hours, not much difference for them: all lost network(not wireless) 2~3 times, all caused reboot one or two times(not sure this is totally wireless related though), 6e18(first four digits in the hash) had one case that failed to self-recover network connection. all download speed is about 6Mbps on average(which is much slower comparing to ath9k STA). For 7daf I got a kernel hang within 10 minutes and I ended up power-cycling the device, also with 7daf my PC/STA can no longer have wireless work, I had to reboot my PC and reload it with "modprobe wl", looks like the STA has some issues too, but I never had problems with other openwrt routers(e.g. tplink) in the last few years while using the same PC.

I will hook a serial cable and do more tests later, the above tests are preliminary. The STA is a ubuntu 18.04 with brcmsmac driver.
